R/unnecessary_anonymous_function_linter.R

```diff
@@ -0,0 +1,94 @@
+#' Block usage of anonymous functions in iteration functions when unnecessary
+#'
+#' Using an anonymous function in, e.g., [lapply()] is not always necessary,
+#'   e.g. `lapply(DF, sum)` is the same as `lapply(DF, function(x) sum(x))` and
+#'   the former is more readable.
+#'
+#' @evalRd rd_tags("unnecessary_anonymous_function_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+unnecessary_anonymous_function_linter <- function() {
+  # TODO(michaelchirico): is there any way to extend this to all (base?) functions
+  #   accepting FUN as an argument? Reduce, Filter, mapply, outer, combn, ...
```

Comment:
``` r
library(parallel)

is_arg_present <- function(.f, arg_name) {
  fun_args <- as.list(formals(.f))
  is_arg_present <- any(names(fun_args) == arg_name)
  return(is_arg_present)
}

funs_with_arg_name <- function(env, arg_name) {
  objs <- mget(ls(env, all = TRUE), inherits = TRUE)
  funs <- Filter(is.function, objs)
  names(purrr::keep(funs, is_arg_present, arg_name = arg_name))
}

purrr::map(
  list("package:base", "package:utils", "package:stats", "package:graphics", "package:methods", "package:parallel"),
  funs_with_arg_name,
  arg_name = "FUN"
) |>
  purrr::simplify()
#>  [1] ".kronecker"           ".mapply"              "apply"               
#>  [4] "by"                   "by.data.frame"        "by.default"          
#>  [7] "eapply"               "kronecker"            "lapply"              
#> [10] "mapply"               "match.fun"            "outer"               
#> [13] "sapply"               "sweep"                "tapply"              
#> [16] "vapply"               "Vectorize"            "combn"               
#> [19] "maphash"              "addmargins"           "aggregate.data.frame"
#> [22] "aggregate.ts"         "ave"                  "dendrapply"          
#> [25] "kronecker"            "mclapply"             "mcmapply"            
#> [28] "parApply"             "parCapply"            "parRapply"           
#> [31] "parSapply"            "parSapplyLB"          "pvec"
```

<sup>Created on 2022-09-20 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
thanks! I think we also need to remove things like `.mapply` that don't support accepting args in `...` form

Reply:
True. Updated the reprex accordingly.

Reply:
I also removed S3 methods and glancing through the manuals, I also exclude `ave()`, `combn()`, `sweep()`, `kronecker()`, and `by()`. With those, the signature is not exactly amenable because `...` should come directly after `FUN` (or be required, like for `vapply()`), otherwise user may have to specify some odd argument in order to add position arguments to the lambda.

That leaves us with these to add for now (as well as any others where the argument is not named `FUN`, but we can get to that later)

```
mapply
outer
parApply
pvec
```

Reply:
Sounds good. 

Thanks for painstakingly going through the docs to decide which ones are appropriate to include here!